### ISSUE TITLE
Hierarchical (mixed-effects) foundation model via amortized VI — VAE-style arm of the hierarchical-Bayes comparison

### ISSUE THREAD

**Issue by hanhou** [AllenNeuralDynamics/aind-dynamic-foraging-bfm-dispatcher#57]
Make the multisubject sequence model an explicit **hierarchical (mixed-effects) random-effects model** — population → subject → session — trained by **amortized variational inference** (VAE-style, no MCMC), so it (a) captures individual *and* session-wise heterogeneity with principled partial pooling, (b) yields per-session posterior **uncertainty**, and (c) lines up **arm-for-arm with the cognitive hierarchical-Bayesian (HB) model** (#20) on the same synthetic ground truth.

📄 **Full design note (committed):** [`docs/design-hierarchical-vi-foundation-model.md`](https://github.com/AllenNeuralDynamics/aind-disrnn-dispatcher/blob/main/docs/design-hierarchical-vi-foundation-model.md) — landed in https://github.com/AllenNeuralDynamics/aind-disrnn-dispatcher/pull/54. **Status: design / plan, not implemented.** This issue tracks building it.

## Why — the study-04 result that motivates this

[`04-gru-vs-disrnn-embedding-recovery`](https://github.com/AllenNeuralDynamics/aind-disrnn-dispatcher/tree/main/studies/04-gru-vs-disrnn-embedding-recovery), variant `gru-stage4b-concentration`, established that **session conditioning is a smooth function of session *position*, not content.** The only session-encoding modes (`none/scalar/fourier`) all feed a delta-MLP from the normalized session *index*. Empirically the session-conditioned embedding predicts session **position** at R² ≈ 0.58–0.66 but adds **zero** over the subject embedding for decoding the realized per-session **family** (gap ≈ 0 at every Dirichlet concentration 0.5 → 5).

So the Stage-4b null is **robust, not a sparsity artifact**: a continuous, position-indexed session code is *structurally unable* to represent a discrete per-session regime draw. Right inductive bias for smooth **drift**; wrong one for discrete **switching**.

## The reframe

The current model is **already a crude two-level mixed-effects model** — shared dynamics (fixed effect) + subject embedding table (subject random effect, L2-shrunk) + session delta (session random effect, `lambda_reg_session`-shrunk, but smoothness-constrained and MAP point-estimated).

> *"The current model **is** this framework with uncertainty and learned variances switched off — we are turning knobs back on, not replacing the model."*

Pooling spectrum: `none` = **complete pooling** across a subject's sessions · `scalar`/`fourier` = **partial pooling** with a smooth position prior · a free/content-inferred per-session latent = **no pooling** (per-session fit). **The third does not exist today. Adding it under a hierarchical prior is the core of this work.**

## Design A (minimal) — hierarchical Gaussian random effects

```
population:  μ                            (learned global mean)
subject:     e_s      ~ N(μ,   Σ_subj)    # subject random effect
session:     e_{s,t}  ~ N(e_s, Σ_sess)    # session deviation from its subject
likelihood:  y_{s,t}  ~ RNN_θ( · | e_{s,t})
```

`Σ_sess / Σ_subj` are the **variance components** — the FM analog of the ICC / random-effect variances an HB reports. Three edits vs today: free sampled session latent (drop the smooth-position constraint); add the `N(e_s, Σ_sess)` shrink-to-subject link; replace fixed `lambda_reg_session` L2 with a *learned* prior (KL term).

## Design B (extension) — discrete regimes
1. **Content-inferred discrete session code** — session encoder → logits over K states via **Gumbel-softmax** (temperature-annealed) or **VQ**. The code index *is* the recovered regime.
2. **Mixture-of-experts** — K expert dynamics + content-based gating; the gate assignment is the per-session regime.

In Bayesian terms these are just a **prior-family choice on the session latent**: Gaussian = graded (A); mixture / Dirichlet-process = discrete regimes (B).

## Arm-for-arm with the cognitive HB (→ #20)

Same statistical object; the latent is hand-specified cognitive params (HB) vs a learned RNN latent (FM). Compare on identical axes, per held-out cell, on shared synthetic ground truth:
- **recovery** — CCA of posterior means to truth (reuse study-04 tooling; the RNN latent is unidentified up to rotation/scale, so score by CCA, not raw coordinates);
- **calibration** — do credible intervals cover the true latent at their stated level? **This is where amortized VI can lose to MCMC-HB at equal predictive LL.**
- `baseline_rl` = correct-model-class reference; `likelihood_relative_to_groundtruth` (ceiling 1.0) = shared currency.

**Held-out is a matrix, not a scalar** — one split per level: trials-within-session · sessions-within-subject (`heldout_session_mode=tail`) · subjects zero-shot (`e = μ`) · subjects few-shot (`auto_heldout_finetune`). **Golden rule:** infer the latent from *context only*, score *disjoint* targets; **marginalize** the posterior rather than plugging in the MAP (scoring at `e = m` overstates fit and unfairly favors the more-confident model).

## Staging (from §8 of the note)
- [ ] **5a — minimal Gaussian hierarchical (Design A).** Learned `Σ`, free session latent, amortized encoder, KL β-warmup. Validate on Stage-1/2 synthetic (recovery + calibration ≥ current point-estimate model), then Stage-2b (drift). **Deliverable: recovery + coverage vs. the existing GRU and vs. cognitive HB.**
- [ ] **5b — discrete session code (Design B, minimal).** Gumbel/VQ; re-run the Stage-4b concentration sweep. **Success = the code index recovers the per-session family where continuous conditioning gave gap ≈ 0.**
- [ ] **5c — mixture-of-experts (Design B, full).** Only if 5b shows a discrete code helps.

## Mapping to existing code (the diff, §6)
- subject table → `q(e_s)`: the current table **plus a log-variance column**.
- [`code/models/session_conditioning.py`](https://github.com/AllenNeuralDynamics/aind-disrnn-wrapper/blob/main/code/models/session_conditioning.py) delta-net → small session encoder (MLP / tiny GRU over the session's trials) emitting `(m, log s²)`; reparam sample; inject at the existing `session_integration_type` point.
- `lambda_reg_session` → `KL(q_sess ‖ N(e_s, Σ_sess))` with **learned** `Σ` (softplus).
- **Reuse the existing λ-forward schedule (`session_curriculum_lambda`) as KL β-annealing** — the must-have against posterior collapse; add free-bits if a dim still collapses.
- Held-out: reuse `heldout_session_mode=tail` + `auto_heldout_finetune`, but swap gradient-fine-tune inference for **encoder-inferred `q` + posterior-predictive scoring**. Metric unchanged.

## Risks (§9)
- **Posterior collapse** — the central VI failure. Needs β-warmup + free-bits + a per-dim KL monitor.
- **Amortization gap** — one shared encoder can't be perfectly tailored per session; close it (semi-amortized / IWAE) only for the final calibration comparison.
- **Is discrete even needed?** Design B is a real architecture change. If within-subject variation is graded (drift/engagement), Stage-2 already covers it and Design A suffices.
- **Identifiability ceiling:** if ~650 trials don't *behaviorally* distinguish the families, no mechanism recovers them — an MoE won't rescue what the data doesn't contain.

## Done when
- Design A implemented and validated on Stage-1/2/2b synthetic: recovery + **coverage** reported vs. the current point-estimate GRU and vs. the cognitive HB (#20), on the shared held-out matrix.
- A go/no-go call on Design B, driven by the 5b Stage-4b re-run.

**Related:** #20 (the HB arm — must share the held-out matrix + metric) · #21 / #16 · parent #33
